bases/renku_data_services/background_jobs/utils.py

```diff
@@ -0,0 +1,137 @@
+"""Background jobs utilities."""
+
+import traceback
+from collections.abc import Coroutine
+from dataclasses import dataclass
+from typing import Self
+
+
+@dataclass
+class BackgroundJobError(Exception):
+    """Error raised when an exception happenend in a background job."""
+
+    message = "Background job failed:"
+    errors: list[BaseException]
+
+    def __format_errors__(self) -> str:
+        """Format contained errors for output."""
+        error_messages = []
+        total_errors = len(self.errors)
+        for i, error in enumerate(self.errors):
+            error_messages.append(
+                f"=== Error {i+1}/{total_errors} ===\n"
+                + "".join(traceback.TracebackException.from_exception(error).format())
+            )
+
+        return "\n".join(error_messages)
+
+    def __repr__(self) -> str:
+        """String representation of the error."""
+        return f"{self.__class__.__qualname__}: {self.message}\n{self.__format_errors__()}"
+
+    def __str__(self) -> str:
+        """String representation of the error."""
+        return f"{self.__class__.__qualname__}: {self.message}\n{self.__format_errors__()}"
+
+
+class ErrorHandlerMonad:
+    """A monad for tracking exceptions of computational steps."""
+
+    def __init__(self, errors: list[BaseException]) -> None:
+        self.errors = errors
+
+    async def map(self, fn: Coroutine[None, None, None] | list[Coroutine[None, None, None]]) -> Self:
+        """Await a single or list of coroutines and track an error if it occurs.
+
+        For a list of coroutines, if any of them fail, subsequent ones aren't executed.
+        """
+        try:
+            if isinstance(fn, list):
+                for f in fn:
+                    await f
+            else:
+                await fn
+        except BaseException as e:
+            self.errors.append(e)
+        return self
+
+    async def flatmap(
+        self, fn: "Coroutine[None,None,ErrorHandlerMonad]|list[Coroutine[None,None,ErrorHandlerMonad]]"
+    ) -> Self:
+        """Await a single or list of coroutines that return ErrorHandlerMonads and track an errors if they occur.
+
+        For a list of coroutines, if any of them fail, subsequent ones aren't executed.
```

Comment:
This doesn't see to be the case. In the `for` loop, the code appends the new errors to `handler`, but that does not stop the loop from executing the subsequent item from `fn`.

Reply:
I guess this is down to what `a coroutine fails` is meant to convey: that it utterly fail or that the returned `ErrorHandlerMonad` contains errors.